dimod/constrained.py

```diff
@@ -408,34 +511,95 @@ def add_discrete(self, variables: Iterable[Variable],
         Returns:
             Label of the added constraint.
 
-        Raises:
-            ValueError: If any of the given variables have already been added
-                to the model with any vartype other than `BINARY`.
+        Examples:
 
-            ValueError: If any of the given variables are already used in
-                another discrete variable.
+            >>> cqm = dimod.ConstrainedQuadraticModel()
+            >>> iterable = ['x', 'y', 'z']
+            >>> for v in iterable:
+            ...      cqm.add_variable(v, 'BINARY')
+            'x'
+            'y'
+            'z'
+            >>> cqm.add_discrete(iterable, label='discrete-xyz')
+            'discrete-xyz'
 
         """
         if label is not None and label in self.constraints:
             raise ValueError("a constraint with that label already exists")
 
-        if isinstance(variables, Iterator):
-            variables = list(variables)
+        bqm = BinaryQuadraticModel(Vartype.BINARY, dtype=np.float32)
 
         for v in variables:
-            if (v in self.variables and
-                    any(v in self.constraints[label].lhs.variables for label in self.discrete)):
-                # the first check is not necessary, but it's faster to make sure it's
-                # somewhere to be found before looking through all of the
-                # discrete constraints
-                raise ValueError(f"variable {v!r} is already used in a discrete variable")
-            elif v in self.variables and self.vartype(v) != Vartype.BINARY:
-                raise ValueError(f"variable {v!r} has already been added but is not BINARY")
-
-        # we can! So add them
-        bqm = BinaryQuadraticModel('BINARY', dtype=np.float32)
-        bqm.add_variables_from((v, 1) for v in variables)
-        label = self.add_constraint(bqm == 1, label=label)
+            if v in self.variables:
+                # it already exists, let's make sure it's not already used
+                if any(v in self.constraints[label].lhs.variables for label in self.discrete):
+                    raise ValueError(f"variable {v!r} is already used in a discrete variable")
+                if self.vartype(v) is not Vartype.BINARY:
+                    raise ValueError(f"variable {v!r} has already been added but is not BINARY")
+
+            bqm.set_linear(v, 1)
+
+        if bqm.num_variables < 2:
+            raise ValueError("discrete constraints must have at least two variables")
+
+        label = self.add_constraint_from_comparison(bqm == 1, label=label, copy=False)
+        self.discrete.add(label)
```

Comment:
This is checked in add_discrete_from_model which is transitively called from here

Reply:
See above